dht.go

```diff
@@ -448,22 +449,37 @@ func (dht *IpfsDHT) putLocal(key string, rec *recpb.Record) error {
 
 // peerFound signals the routingTable that we've found a peer that
 // might support the DHT protocol.
+// If we have a connection a peer but no exchange of a query RPC ->
+//    LastQueriedAt=time.Now (so we don't ping it for some time for a liveliness check)
+//    LastUsefulAt=N/A
+// If we connect to a peer and exchange a query RPC ->
+//    LastQueriedAt=time.Now (same reason as above)
+//    LastUsefulAt=time.Now (so we give it some life in the RT without immediately evicting it)
+// If we query a peer we already have in our Routing Table ->
+//    LastQueriedAt=time.Now()
+//    LastUsefulAt remains unchanged
+// If we connect to a peer we already have in the RT but do not exchange a query (rare)
+//    Do Nothing.
 func (dht *IpfsDHT) peerFound(ctx context.Context, p peer.ID, queryPeer bool) {
 	logger.Debugw("peer found", "peer", p)
 	b, err := dht.validRTPeer(p)
 	if err != nil {
 		logger.Errorw("failed to validate if peer is a DHT peer", "peer", p, "error", err)
 	} else if b {
-		_, err := dht.routingTable.TryAddPeer(p, queryPeer)
+		newlyAdded, err := dht.routingTable.TryAddPeer(p, queryPeer)
 		if err != nil {
 			// peer not added.
 			return
 		}
 
-		// If we discovered the peer because of a query, we need to ensure we override the "zero" lastSuccessfulOutboundQuery
+		// If we freshly added the peer because of a query, we need to ensure we override the "zero" lastUsefulAt
 		// value that must have been set in the Routing Table for this peer when it was first added during a connection.
-		if queryPeer {
-			dht.routingTable.UpdateLastSuccessfulOutboundQuery(p, time.Now())
+		if newlyAdded && queryPeer {
+			dht.routingTable.UpdateLastUsefulAt(p, time.Now())
```

Comment:
If this _is_ a new peer, don't set this to `now` in `addPeer` regardless?

Reply:
@aarshkshah1992 not blocking a merge on this as it shouldn't really matter. We can fix this later.

Reply:
@Stebalien The idea is that if we discover and add a peer because of an inbound/outbound query, we do NOT want to evict the peer for one ''cycle".

Note that the time will NOT be "now" for peers we connect to without querying them .

Reply:
Ah! Thanks!

Reply:
Talking this over with @aschmahmann, I think I was right. In `addPeer` (in the routing table), [if `queryPeer` is true](https://github.com/libp2p/go-libp2p-kbucket/blob/master/table.go#L114-L116), we'll set "last useful" to now _when_ the peer wasn't already in our routing table. Then, when the peer wasn't already in our routing table, we'll mark them as useful again.

Reply:
@Stebalien I see what you mean. I am removing this particular check from here and adding [this commit](https://github.com/libp2p/go-libp2p-kbucket/pull/77/commits/9ee07aec44d41f9e4b399568289c00e4354d482b) so when we first query a peer after connecting to it(that's two calls to `addPeer`, once on successful query and once on connection), we give it one and ONLY one usefulness bump.